File: packages/vault-core/tests/binding.test.ts
import { describe, expect, it } from "vitest";
import { createClock } from "../src/hlc.js";
import { MemoryVaultStorage, VAULT_DOCUMENT } from "../src/storage.js";
import { VaultSync } from "../src/sync.js";
import { FakeRemote } from "../src/testing.js";
import { datasetOf, fingerprint, itemField, visibleItems } from "../src/model.js";

/**
 * Moving a keyring into its own document, and back out of it.
 *
 * This is the step that turns a private keyring into a shareable one, and it
 * is the only place in the app where passwords are deliberately taken out of
 * one file and put into another. Every failure here is a lost password, so
 * the tests care mostly about what is true *between* the two writes.
 */
function engine() {
  const storage = new MemoryVaultStorage();
  const vaultRemote = new FakeRemote();
  const datasetRemote = new FakeRemote();
  const sync = new VaultSync({
    storage,
    remote: vaultRemote,
    remoteFor: (documentId) =>
      documentId === VAULT_DOCUMENT
        ? vaultRemote
        : documentId === "ds-house"
          ? datasetRemote
          : null,
    clock: createClock({ node: "test" }),
  });
  return { storage, sync, vaultRemote, datasetRemote };
}

async function withHousehold() {
  const parts = engine();
  await parts.sync.putKeyring({ keyringId: "personal", name: "Just mine" });
  await parts.sync.putKeyring({ keyringId: "house", name: "Household" });
  await parts.sync.putItem({ itemId: "bank", keyringId: "personal", fields: { title: "Bank" } });
  await parts.sync.putItem({
    itemId: "wifi",
    keyringId: "house",
    fields: { title: "Wifi", password: "hunter2" },
  });
  await parts.sync.putItem({ itemId: "gas", keyringId: "house", fields: { title: "Gas" } });
  return parts;
}

describe("binding a keyring to its own document", () => {
  it("moves its passwords out of the vault and nothing else", async () => {
    const { storage, sync } = await withHousehold();
    await sync.bindKeyring("house", "ds-house");

    const vault = await storage.readState(VAULT_DOCUMENT);
    const dataset = await storage.readState("ds-house");

    expect(Object.keys(dataset.items).sort()).toEqual(["gas", "wifi"]);
    expect(itemField(dataset.items["wifi"]!, "password")).toBe("hunter2");

    // What stays behind in the vault is a scrubbed tombstone, not the
    // password. A record has to stay — a CRDT has no way to say "gone" except
    // by saying it — but it carries no value any more.
    expect(vault.items["wifi"]?.deleted.value).toBe(true);
    expect(itemField(vault.items["wifi"]!, "password")).toBe("");
    expect(vault.items["wifi"]?.history).toEqual([]);
    expect(vault.items["gas"]?.deleted.value).toBe(true);

    // The keyring that was not shared is untouched, in the file it was in.
    expect(vault.items["bank"]?.deleted.value).toBe(false);
  });

  it("keeps the vault able to name what it no longer holds", async () => {
    const { storage, sync } = await withHousehold();
    await sync.bindKeyring("house", "ds-house");

    const vault = await storage.readState(VAULT_DOCUMENT);
    expect(vault.keyrings["house"]?.name.value).toBe("Household");
    expect(datasetOf(vault.keyrings["house"])).toBe("ds-house");
  });

  it("changes nothing a person can see", async () => {
    const { sync } = await withHousehold();
    const before = visibleItems(await sync.state())
      .map((item) => item.id)
      .sort();

    await sync.bindKeyring("house", "ds-house");

    const after = visibleItems(await sync.state())
      .map((item) => item.id)
      .sort();
    expect(after).toEqual(before);
    const wifi = (await sync.state()).items["wifi"];
    expect(itemField(wifi!, "password")).toBe("hunter2");
  });

  /**
   * The crash-safety property the ordering exists for. If the process dies
   * between the two writes, the items must still be somewhere — and the only
   * way to guarantee that is to write the copy before removing the original.
   */
  it("never has the passwords in neither document", async () => {
    const { storage, sync } = await withHousehold();

    const seen: string[] = [];
    const original = storage.commitAll.bind(storage);
    let interrupted = false;
    storage.commitAll = async (ops, next, documentId = VAULT_DOCUMENT) => {
      // Refuse the second write, exactly as a crash would.
      if (interrupted) throw new Error("crashed");
      seen.push(documentId);
      if (seen.length === 1) {
        await original(ops, next, documentId);
        interrupted = true;
        return;
      }
      await original(ops, next, documentId);
    };

    await expect(sync.bindKeyring("house", "ds-house")).rejects.toThrow("crashed");

    // The dataset was written first, so the passwords survived.
    expect(seen[0]).toBe("ds-house");
    const dataset = await storage.readState("ds-house");
    expect(Object.keys(dataset.items).sort()).toEqual(["gas", "wifi"]);
    // And the vault still has them too, because the removal never happened.
    const vault = await storage.readState(VAULT_DOCUMENT);
    expect(vault.items["wifi"]).toBeDefined();
  });

  it("refuses to bind a keyring twice", async () => {
    const { sync } = await withHousehold();
    await sync.bindKeyring("house", "ds-house");
    await expect(sync.bindKeyring("house", "ds-other")).rejects.toThrow(/already lives/);
  });

  it("is a no-op when asked for the binding it already has", async () => {
    const { sync } = await withHousehold();
    await sync.bindKeyring("house", "ds-house");
    await expect(sync.bindKeyring("house", "ds-house")).resolves.toBeDefined();
  });

  /**
   * The bug this caught: binding changes nothing else in the vault, so a
   * fingerprint that ignored the binding reported "unchanged" and the sync
   * skipped the upload. Other devices would then never learn the keyring had
   * moved, and would keep looking for its passwords in the vault.
   */
  it("changes the vault's fingerprint, so it gets published", async () => {
    const { storage, sync, vaultRemote } = await withHousehold();
    await sync.sync();
    const before = fingerprint(await storage.readState(VAULT_DOCUMENT));

    await sync.bindKeyring("house", "ds-house");
    const after = fingerprint(await storage.readState(VAULT_DOCUMENT));
    expect(after).not.toBe(before);

    const writes = vaultRemote.writes;
    await sync.sync();
    expect(vaultRemote.writes).toBeGreaterThan(writes);
    expect(datasetOf(vaultRemote.snapshot().keyrings["house"])).toBe("ds-house");
  });

  it("publishes the keyring's passwords to its own remote and no other", async () => {
    const { sync, vaultRemote, datasetRemote } = await withHousehold();
    await sync.bindKeyring("house", "ds-house");
    await sync.sync();

    expect(Object.keys(datasetRemote.snapshot().items).sort()).toEqual(["gas", "wifi"]);
    expect(itemField(datasetRemote.snapshot().items["wifi"]!, "password")).toBe("hunter2");
    // The backup of the private vault keeps no copy of the shared password.
    expect(itemField(vaultRemote.snapshot().items["wifi"]!, "password")).toBe("");
    expect(vaultRemote.snapshot().items["bank"]?.deleted.value).toBe(false);
  });

  /** A rename is part of what was shared, so it has to travel with the items. */
  it("sends a later rename to the shared document", async () => {
    const { storage, sync } = await withHousehold();
    await sync.bindKeyring("house", "ds-house");
    await sync.putKeyring({ keyringId: "house", name: "Ours" });

    const dataset = await storage.readState("ds-house");
    expect(dataset.keyrings["house"]?.name.value).toBe("Ours");
    // And it is what the person sees, whichever document it came from.
    expect((await sync.state()).keyrings["house"]?.name.value).toBe("Ours");
  });
});

describe("unbinding a keyring", () => {
  it("brings the passwords back into the vault", async () => {
    const { storage, sync } = await withHousehold();
    await sync.bindKeyring("house", "ds-house");
    await sync.unbindKeyring("house");

    const vault = await storage.readState(VAULT_DOCUMENT);
    expect(Object.keys(vault.items).sort()).toEqual(["bank", "gas", "wifi"]);
    expect(datasetOf(vault.keyrings["house"])).toBeNull();
  });

  it("keeps edits made while it was shared", async () => {
    const { sync } = await withHousehold();
    await sync.bindKeyring("house", "ds-house");
    await sync.putItem({ itemId: "wifi", keyringId: "house", fields: { password: "rotated" } });
    await sync.unbindKeyring("house");

    const wifi = (await sync.state()).items["wifi"];
    expect(itemField(wifi!, "password")).toBe("rotated");
  });

  it("stops syncing the document it no longer uses", async () => {
    const { sync, datasetRemote } = await withHousehold();
    await sync.bindKeyring("house", "ds-house");
    await sync.sync();
    const writes = datasetRemote.writes;

    await sync.unbindKeyring("house");
    await sync.putItem({ itemId: "gas", keyringId: "house", fields: { title: "Gas co" } });
    await sync.sync();

    expect(datasetRemote.writes).toBe(writes);
  });
});

describe("leaving a keyring someone else shared", () => {
  it("hides it here without deleting anyone else's passwords", async () => {
    const { storage, sync, datasetRemote } = await withHousehold();
    await sync.bindKeyring("house", "ds-house");
    await sync.sync();
    const theirs = datasetRemote.snapshot();

    await sync.leaveKeyring("house");

    // Gone from this vault...
    expect(visibleItems(await sync.state()).map((item) => item.id)).toEqual(["bank"]);
    // ...and the shared document is untouched, tombstone-free.
    await sync.sync();
    expect(Object.keys(datasetRemote.snapshot().items).sort()).toEqual(["gas", "wifi"]);
    expect(datasetRemote.snapshot().items["wifi"]?.deleted.value).toBe(false);
    expect(theirs.items["wifi"]?.deleted.value).toBe(false);

    // The tombstone is local: it lives in the vault, not in what was shared.
    const vault = await storage.readState(VAULT_DOCUMENT);
    expect(vault.keyrings["house"]?.deleted.value).toBe(true);
    expect(datasetRemote.snapshot().keyrings["house"]?.deleted.value).toBe(false);
  });
});

describe("moving a password across documents", () => {
  /**
   * The property the purge exists for. "I took that one out of the shared
   * folder" has to be true of the file the other person reads, not just of
   * this screen.
   */
  it("takes it away from the shared document", async () => {
    const { sync, datasetRemote } = await withHousehold();
    await sync.bindKeyring("house", "ds-house");
    await sync.sync();
    expect(itemField(datasetRemote.snapshot().items["wifi"]!, "password")).toBe("hunter2");

    await sync.moveItem("wifi", "personal");
    await sync.sync();

    const shared = datasetRemote.snapshot().items["wifi"];
    expect(shared?.deleted.value).toBe(true);
    expect(itemField(shared!, "password")).toBe("");
    expect(shared?.history).toEqual([]);
  });

  it("keeps it readable in the keyring it moved to", async () => {
    const { sync } = await withHousehold();
    await sync.bindKeyring("house", "ds-house");
    await sync.moveItem("wifi", "personal");

    const state = await sync.state();
    expect(state.items["wifi"]?.deleted.value).toBe(false);
    expect(state.items["wifi"]?.keyring.value).toBe("personal");
    expect(itemField(state.items["wifi"]!, "password")).toBe("hunter2");
    expect(visibleItems(state).map((item) => item.id).sort()).toEqual(["bank", "gas", "wifi"]);
  });

  it("is an ordinary move when both keyrings are in the same document", async () => {
    const { storage, sync } = await withHousehold();
    await sync.putKeyring({ keyringId: "work", name: "Work" });
    await sync.moveItem("bank", "work");

    const vault = await storage.readState(VAULT_DOCUMENT);
    // Not relocated: the record keeps its history and its values in place.
    expect(vault.items["bank"]?.keyring.value).toBe("work");
    expect(vault.items["bank"]?.deleted.value).toBe(false);
  });
});

File: packages/android/vault/src/test/kotlin/app/keyweb/vault/BindingTest.kt
package app.keyweb.vault

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue
import kotlinx.coroutines.test.runTest

/**
 * Moving a keyring into its own document, and back out of it.
 *
 * This is the step that turns a private keyring into a shareable one, and the
 * only place in the app where passwords are deliberately taken out of one file
 * and put into another. Every failure here is a lost password, so these tests
 * care mostly about what is true *between* the two writes.
 *
 * Parity with the web's `binding.test.ts`, question for question.
 */
class BindingTest {

    private class Rig {
        val storage = MemoryVaultStorage()
        val vaultRemote = FakeRemote()
        val datasetRemote = FakeRemote()
        val sync = VaultSync(
            storage = storage,
            remote = vaultRemote,
            remoteFor = { documentId ->
                when (documentId) {
                    VAULT_DOCUMENT -> vaultRemote
                    "ds-house" -> datasetRemote
                    else -> null
                }
            },
            clock = Clock("test"),
        )
    }

    private suspend fun household(): Rig {
        val rig = Rig()
        rig.sync.putKeyring(keyringId = "personal", name = "Just mine")
        rig.sync.putKeyring(keyringId = "house", name = "Household")
        rig.sync.putItem(itemId = "bank", keyringId = "personal", fields = mapOf("title" to "Bank"))
        rig.sync.putItem(
            itemId = "wifi",
            keyringId = "house",
            fields = mapOf("title" to "Wifi", "password" to "hunter2"),
        )
        rig.sync.putItem(itemId = "gas", keyringId = "house", fields = mapOf("title" to "Gas"))
        return rig
    }

    @Test
    fun `moves its passwords out of the vault and nothing else`() = runTest {
        val rig = household()
        rig.sync.bindKeyring("house", "ds-house")

        val vault = rig.storage.readState(VAULT_DOCUMENT)
        val dataset = rig.storage.readState("ds-house")

        assertEquals(listOf("gas", "wifi"), dataset.items.keys.sorted())
        assertEquals("hunter2", dataset.items.getValue("wifi").field("password"))

        // What stays behind in the vault is a scrubbed tombstone, not the
        // password. A record has to stay — a CRDT has no way to say "gone"
        // except by saying it — but it carries no value any more.
        assertTrue(vault.items.getValue("wifi").deleted.value)
        assertEquals("", vault.items.getValue("wifi").field("password"))
        assertTrue(vault.items.getValue("wifi").history.isEmpty())
        assertTrue(vault.items.getValue("gas").deleted.value)

        // The keyring that was not shared is untouched, in the file it was in.
        assertTrue(!vault.items.getValue("bank").deleted.value)
    }

    @Test
    fun `keeps the vault able to name what it no longer holds`() = runTest {
        val rig = household()
        rig.sync.bindKeyring("house", "ds-house")

        val vault = rig.storage.readState(VAULT_DOCUMENT)
        assertEquals("Household", vault.keyrings.getValue("house").name.value)
        assertEquals("ds-house", datasetOf(vault.keyrings["house"]))
    }

    @Test
    fun `changes nothing a person can see`() = runTest {
        val rig = household()
        val before = visibleItems(rig.sync.state()).map { it.id }.sorted()

        rig.sync.bindKeyring("house", "ds-house")

        val after = rig.sync.state()
        assertEquals(before, visibleItems(after).map { it.id }.sorted())
        assertEquals("hunter2", after.items.getValue("wifi").field("password"))
    }

    /**
     * The crash-safety property the ordering exists for. If the process dies
     * between the two writes, the passwords must still be somewhere — and the
     * only way to guarantee that is to write the copy before removing the
     * original.
     */
    @Test
    fun `never has the passwords in neither document`() = runTest {
        val rig = Rig()
        val seen = mutableListOf<String>()
        val failing = object : VaultStorage by rig.storage {
            override suspend fun commitAll(
                ops: List<VaultOp>,
                nextState: VaultState,
                documentId: String,
            ) {
                // Refuse the second write, exactly as a crash would.
                if (seen.isNotEmpty()) error("crashed")
                seen += documentId
                rig.storage.commitAll(ops, nextState, documentId)
            }
        }
        val sync = VaultSync(storage = failing, remote = rig.vaultRemote, clock = Clock("test"))
        rig.storage.commitAll(
            listOf(
                VaultOp.KeyringPut("k", "001700000000000-00000-test", "house", "Household"),
                VaultOp.ItemPut(
                    "i",
                    "001700000000000-00001-test",
                    "wifi",
                    "house",
                    mapOf("password" to "hunter2"),
                ),
            ),
            applyOps(
                emptyVault(),
                listOf(
                    VaultOp.KeyringPut("k", "001700000000000-00000-test", "house", "Household"),
                    VaultOp.ItemPut(
                        "i",
                        "001700000000001-00001-test",
                        "wifi",
                        "house",
                        mapOf("password" to "hunter2"),
                    ),
                ),
            ),
            VAULT_DOCUMENT,
        )

        assertFailsWith<IllegalStateException> { sync.bindKeyring("house", "ds-house") }

        // The dataset was written first, so the passwords survived.
        assertEquals(listOf("ds-house"), seen)
        assertEquals("hunter2", rig.storage.readState("ds-house").items.getValue("wifi").field("password"))
        // And the vault still has them too, because the removal never happened.
        assertEquals(
            "hunter2",
            rig.storage.readState(VAULT_DOCUMENT).items.getValue("wifi").field("password"),
        )
    }

    @Test
    fun `refuses to bind a keyring twice`() = runTest {
        val rig = household()
        rig.sync.bindKeyring("house", "ds-house")
        assertFailsWith<IllegalStateException> { rig.sync.bindKeyring("house", "ds-other") }
    }

    /**
     * The bug this caught: binding changes nothing else in the vault, so a
     * fingerprint that ignored the binding reported "unchanged" and the sync
     * skipped the upload. Other devices would then never learn the keyring had
     * moved, and would keep looking for its passwords in the vault.
     */
    @Test
    fun `changes the vault's fingerprint, so it gets published`() = runTest {
        val rig = household()
        rig.sync.sync()
        val before = fingerprint(rig.storage.readState(VAULT_DOCUMENT))

        rig.sync.bindKeyring("house", "ds-house")
        assertTrue(fingerprint(rig.storage.readState(VAULT_DOCUMENT)) != before)

        val writes = rig.vaultRemote.writes
        rig.sync.sync()
        assertTrue(rig.vaultRemote.writes > writes)
        assertEquals("ds-house", datasetOf(rig.vaultRemote.snapshot().keyrings["house"]))
    }

    @Test
    fun `publishes the keyring's passwords to its own remote and no other`() = runTest {
        val rig = household()
        rig.sync.bindKeyring("house", "ds-house")
        rig.sync.sync()

        assertEquals(listOf("gas", "wifi"), rig.datasetRemote.snapshot().items.keys.sorted())
        assertEquals("hunter2", rig.datasetRemote.snapshot().items.getValue("wifi").field("password"))
        // The backup of the private vault keeps no copy of the shared password.
        assertEquals("", rig.vaultRemote.snapshot().items.getValue("wifi").field("password"))
        assertTrue(!rig.vaultRemote.snapshot().items.getValue("bank").deleted.value)
    }

    /** A rename is part of what was shared, so it has to travel with the items. */
    @Test
    fun `sends a later rename to the shared document`() = runTest {
        val rig = household()
        rig.sync.bindKeyring("house", "ds-house")
        rig.sync.putKeyring(keyringId = "house", name = "Ours")

        assertEquals("Ours", rig.storage.readState("ds-house").keyrings.getValue("house").name.value)
        assertEquals("Ours", rig.sync.state().keyrings.getValue("house").name.value)
    }

    @Test
    fun `unbinding brings the passwords back into the vault`() = runTest {
        val rig = household()
        rig.sync.bindKeyring("house", "ds-house")
        rig.sync.unbindKeyring("house")

        val vault = rig.storage.readState(VAULT_DOCUMENT)
        assertEquals("hunter2", vault.items.getValue("wifi").field("password"))
        assertTrue(!vault.items.getValue("wifi").deleted.value)
        assertNull(datasetOf(vault.keyrings["house"]))
    }

    @Test
    fun `leaving hides it here without deleting anyone else's passwords`() = runTest {
        val rig = household()
        rig.sync.bindKeyring("house", "ds-house")
        rig.sync.sync()

        rig.sync.leaveKeyring("house")

        assertEquals(listOf("bank"), visibleItems(rig.sync.state()).map { it.id })

        rig.sync.sync()
        val shared = rig.datasetRemote.snapshot()
        assertEquals("hunter2", shared.items.getValue("wifi").field("password"))
        assertTrue(!shared.items.getValue("wifi").deleted.value)
        assertTrue(!shared.keyrings.getValue("house").deleted.value)

        // The tombstone is local: it lives in the vault, not in what was shared.
        assertTrue(rig.storage.readState(VAULT_DOCUMENT).keyrings.getValue("house").deleted.value)
    }

    @Test
    fun `moving a password out of a shared keyring takes it away`() = runTest {
        val rig = household()
        rig.sync.bindKeyring("house", "ds-house")
        rig.sync.sync()
        assertEquals("hunter2", rig.datasetRemote.snapshot().items.getValue("wifi").field("password"))

        rig.sync.moveItem("wifi", "personal")
        rig.sync.sync()

        val shared = assertNotNull(rig.datasetRemote.snapshot().items["wifi"])
        assertTrue(shared.deleted.value)
        assertEquals("", shared.field("password"))
        assertTrue(shared.history.isEmpty())

        // And it is still readable where it moved to.
        val state = rig.sync.state()
        assertEquals("personal", state.items.getValue("wifi").keyring.value)
        assertEquals("hunter2", state.items.getValue("wifi").field("password"))
    }
}
